pkg/infrastructure/azure/azure.go

```diff
@@ -1,14 +1,341 @@
 package azure
 
 import (
+	"context"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v4"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/msi/armmsi"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/storage/armstorage"
+	"github.com/coreos/stream-metadata-go/arch"
+	"github.com/sirupsen/logrus"
+	"k8s.io/utils/ptr"
+
 	"github.com/openshift/installer/pkg/infrastructure/clusterapi"
-	azuretypes "github.com/openshift/installer/pkg/types/azure"
+	"github.com/openshift/installer/pkg/rhcos"
+	aztypes "github.com/openshift/installer/pkg/types/azure"
 )
 
-var _ clusterapi.Provider = (*Provider)(nil)
-
 // Provider implements Azure CAPI installation.
-type Provider struct{}
+type Provider struct {
+	clusterapi.InfraProvider
+	ResourceGroupName    string
+	StorageAccountName   string
+	StorageURL           string
+	StorageAccount       *armstorage.Account
+	StorageClientFactory *armstorage.ClientFactory
+	StorageAccountKeys   []armstorage.AccountKey
+	Tags                 map[string]*string
+}
+
+var _ clusterapi.PreProvider = (*Provider)(nil)
+var _ clusterapi.InfraReadyProvider = (*Provider)(nil)
+
+// Name returns the name of the provider.
+func (p *Provider) Name() string {
+	return aztypes.Name
+}
+
+// PreProvision is called before provisioning using CAPI controllers has begun.
+func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
+	session, err := in.InstallConfig.Azure.Session()
+	if err != nil {
+		return fmt.Errorf("failed to get session: %w", err)
+	}
+
+	installConfig := in.InstallConfig.Config
+	platform := installConfig.Platform.Azure
+	subscriptionID := session.Credentials.SubscriptionID
+	cloudConfiguration := session.CloudConfig
+	tokenCredential := session.TokenCreds
+	resourceGroupName := platform.ClusterResourceGroupName(in.InfraID)
+
+	userTags := platform.UserTags
+	tags := make(map[string]*string, len(userTags)+1)
+	tags[fmt.Sprintf("kubernetes.io_cluster.%s", in.InfraID)] = ptr.To("owned")
+	for k, v := range userTags {
+		tags[k] = ptr.To(v)
+	}
+
+	// Create resource group
+	resourcesClientFactory, err := armresources.NewClientFactory(
```

Comment:
We need to use existing resource group if it's mentioned in the install config. I can handle it in another PR.

Reply:
Patrick & I discussed this, and we will be creating the resource group in PreProvision so that we can set ManagedBy.

Reply:
Makes sense but I was talking about the installer support to install the cluster resources in an already created resource group for which we need to forego creating the resource group.